lib/Server.js

```diff
@@ -329,251 +335,299 @@ class Server {
+          // byPass to that url
+          req.url = bypassUrl;
+          next();
+        } else if (proxyMiddleware) {
+          return proxyMiddleware(req, res, next);
+        } else {
+          next();
+        }
+      });
+    });
+  }
+
+  setupHistoryApiFallbackFeature() {
+    const fallback =
+      typeof this.options.historyApiFallback === 'object'
+        ? this.options.historyApiFallback
+        : null;
+
+    // Fall back to /index.html if nothing else matches.
+    this.app.use(historyApiFallback(fallback));
+  }
+
+  setupStaticFeature() {
+    const contentBase = this.options.contentBase;
+
+    if (Array.isArray(contentBase)) {
+      contentBase.forEach((item) => {
+        this.app.get('*', express.static(item));
+      });
+    } else if (/^(https?:)?\/\//.test(contentBase)) {
+      this.log.warn(
+        'Using a URL as contentBase is deprecated and will be removed in the next major version. Please use the proxy option instead.'
+      );
 
+      this.log.warn(
+        'proxy: {\n\t"*": "<your current contentBase configuration>"\n}'
+      );
+
+      // Redirect every request to contentBase
+      this.app.get('*', (req, res) => {
+        res.writeHead(302, {
+          Location: contentBase + req.path + (req._parsedUrl.search || ''),
+        });
+
+        res.end();
+      });
+    } else if (typeof contentBase === 'number') {
+      this.log.warn(
+        'Using a number as contentBase is deprecated and will be removed in the next major version. Please use the proxy option instead.'
+      );
+
+      this.log.warn(
+        'proxy: {\n\t"*": "//localhost:<your current contentBase configuration>"\n}'
+      );
+
+      // Redirect every request to the port contentBase
+      this.app.get('*', (req, res) => {
+        res.writeHead(302, {
+          Location: `//localhost:${contentBase}${req.path}${req._parsedUrl
+            .search || ''}`,
+        });
+
+        res.end();
+      });
+    } else {
+      // route content request
+      this.app.get(
+        '*',
+        express.static(contentBase, this.options.staticOptions)
+      );
+    }
+  }
+
+  setupServeIndexFeature() {
+    const contentBase = this.options.contentBase;
+
+    if (Array.isArray(contentBase)) {
+      contentBase.forEach((item) => {
+        this.app.get('*', serveIndex(item));
+      });
+    } else if (
+      !/^(https?:)?\/\//.test(contentBase) &&
+      typeof contentBase !== 'number'
+    ) {
+      this.app.get('*', serveIndex(contentBase));
+    }
+  }
+
+  setupWatchStaticFeature() {
+    const contentBase = this.options.contentBase;
+
+    if (
+      /^(https?:)?\/\//.test(contentBase) ||
+      typeof contentBase === 'number'
+    ) {
+      throw new Error('Watching remote files is not supported.');
+    } else if (Array.isArray(contentBase)) {
+      contentBase.forEach((item) => {
+        this._watch(item);
+      });
+    } else {
+      this._watch(contentBase);
+    }
```

Comment:
I see a bug here, when you use string we check on remove, but if you use array we don't check this :trollface: Let's fix it in other PR

Reply:
Can you create an issue? (or already existed?)

Reply:
@hiroppy done https://github.com/webpack/webpack-dev-server/issues/1879